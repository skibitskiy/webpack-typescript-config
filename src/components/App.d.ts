declare type comment = {
    author: string;
    text: string;
};
