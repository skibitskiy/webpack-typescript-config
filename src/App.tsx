import React, { useState } from 'react';

function App() {
    const [multipliers, setMultipliers] = useState([0, 0]);

    const onChangeHandler = ({ target }) => {
        const { index } = target.dataset;
        const { value } = target;
        const res = [...multipliers];
        res[index] = value;
        setMultipliers(res);
    }

    return (
        <div className='app'>
            <div>
            {
                multipliers.map((value, index) =>
                    <input data-index={index} value={value} onChange={onChangeHandler}/>
                )
            }
            </div>
            <div className='answer'>
                { multipliers.reduce((res, next) => res * next) }
            </div>
        </div>
    )
}

export default App;
