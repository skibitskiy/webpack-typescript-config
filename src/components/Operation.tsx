import React, { useState, useCallback } from 'react';
import {
    StyledUserInput,
    StyledResult,
    StyledOperation
} from './styledApp.ts';

export interface OperationProps {
    operation: Function;
    title: string;
}

const Operation: React.FC<OperationProps> = ({operation, title}) => {
    const [nums, setNums] = useState(['0', '0']);
    const cleanNums: Array<number> = nums.map((num: string) => {
        const res = parseInt(num);
        return isNaN(res) ? 0 : res;
    });
    const result: number = operation(cleanNums);

    const onChangeHandler = useCallback(({ target }) => {
        const { index }: { index: number } = target.dataset;
        const { value }: {value: string} = target;
        
        setNums((state) => {
            const res: Array<string> = [...state];
            res[index] = value;
            return res;
        });
    }, []);

    return (
        <StyledOperation onChange={onChangeHandler}>
            <h2>{title}</h2>
            <div>
                { nums.map((value, index) => <StyledUserInput data-index={index} value={value}/>) }
            </div>
            <StyledResult>
                Ответ: { isNaN(result) ? '--' : result }
            </StyledResult>
        </StyledOperation>
    )
}

export default React.memo(Operation);
