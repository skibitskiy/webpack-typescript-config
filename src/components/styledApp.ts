import styled from 'styled-components';

const StyledApp = styled.div`
    height: 100vh;

    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    
    color: ${(props) => props.theme.fontColor};
    background-color: ${(props) => props.theme.mainBgColor};
`;

const StyledUserInput = styled.input`
    height: 1em;
    width: 2em;

    font-size: 48px;
    color: ${(props) => props.theme.fontColor};
    background-color: ${(props) => props.theme.elementBgColor};
    border: 2px solid rgba(0, 0, 0, 0.5);
    border-radius: 5px;
    outline: none;

    & + & {
        margin-left: 16px;
    }
`;

const StyledResult = styled.div`
    font-size: 48px;
`;

const StyledOperation = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;

    & + & {
        margin-top: 24px;
    }
`;

const StyledButton = styled.button`
    padding: 8px;
    
    outline: 0;
    border: 2px solid rgba(0, 0, 0, 0.5);
    border-radius: 5px;
    color: ${(props) => props.theme.fontColor};
    background-color: ${(props) => props.theme.elementBgColor}
`;

export {
    StyledApp,
    StyledUserInput,
    StyledResult,
    StyledOperation,
    StyledButton
}