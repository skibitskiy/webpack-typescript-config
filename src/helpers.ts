/// <reference path="helpers.d.ts" />

const themes: { day: theme, night: theme } = {
    day: {
        fontColor: '#000',
        mainBgColor: '#f0f0f0',
        elementBgColor: '#e8e8e8'
    },
    night: {
        fontColor: '#9fa8da',
        mainBgColor: '#000051',
        elementBgColor: '#534bae'
    }
}

function multiply(nums: Array<number>): number {
    return nums.reduce((res, next) => res * next)
}

function sum(nums: Array<number>): number {
    return nums.reduce((res, next) => res + next)
}

export {
    themes,
    multiply,
    sum
}
