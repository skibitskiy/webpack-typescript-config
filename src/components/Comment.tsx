/// <reference path='App.d.ts' />
import React from 'react'
import { StyledCard, StyledCommentHeader } from './styledApp.ts';

const Comment: React.FC<comment> = ({ author, text }) => {
    return (
        <StyledCard>
            <StyledCommentHeader>
                { author }
            </StyledCommentHeader>
            { text }
        </StyledCard>
    )
}

export default Comment;
