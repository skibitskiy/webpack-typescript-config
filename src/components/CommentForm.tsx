import React, { useState, useCallback } from 'react';
import { StyledCard, StyledCommentHeader, StyledTextarea, StyledButton } from './styledApp.ts';

const CommentForm: React.FC<{ addComment: Function }> = ({ addComment }) => {
    const [text, setText] = useState('');
    const updateComment = useCallback((event) => {
        setText(event.target.value);
    }, [])
    const submitComment = useCallback(() => {
        addComment({ text, author: 'default' });
        setText('');
    }, [text]);

    return (
        <StyledCard>
            <StyledCommentHeader>Comment</StyledCommentHeader>
            <StyledTextarea onChange={updateComment} placeholder='comment...' value={text}/>
            <StyledButton onClick={submitComment}>Отравить</StyledButton>
        </StyledCard>
    )
}

export default CommentForm;
