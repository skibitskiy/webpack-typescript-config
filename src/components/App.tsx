import React from 'react';
import { StyledApp } from './styledApp.ts';
import Operation from './Operation.tsx'
import { multiply, sum } from '../helpers.ts';

function App() {
    return (
        <StyledApp>
            <Operation operation={multiply} title={'Multiply'}/>
            <Operation operation={sum} title={'Summarize'}/>
        </StyledApp>
    )
}

export default App;
