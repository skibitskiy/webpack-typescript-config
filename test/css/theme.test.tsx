import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import initStoryshots from '@storybook/addon-storyshots';
import { create } from 'react-test-renderer';
import { render, act, cleanup } from '@testing-library/react';
import user from '@testing-library/user-event';

import { StyledApp, StyledUserInput, StyledButton } from '../../src/components/styledApp';
import App from '../../src/components/App.tsx';
import { themes } from '../../src/helpers.ts';

initStoryshots();

const assertStyles = (element: Element, styles) => {
    Object.entries(styles).forEach(([rule, value]) => {
        expect(element).toHaveStyleRule(rule, value);
    });
};

describe('Styled-Components', () => {
    afterEach(cleanup);

    it('StyledApp', () => {
        const doc = create(
            <StyledApp theme={themes.day}/>
        ).toJSON();
        expect(doc).toHaveStyleRule('color', themes.day.fontColor);
        expect(doc).toHaveStyleRule('background-color', themes.day.mainBgColor);
    });
    
    it('StyledUserInput', () => {
        const doc = create(
            <StyledUserInput theme={themes.day}/>
        ).toJSON();
        expect(doc).toHaveStyleRule('color', themes.day.fontColor);
        expect(doc).toHaveStyleRule('background-color', themes.day.elementBgColor);
    });

    it('StyledButton', () => {
        const doc = create(
            <StyledButton theme={themes.day} />
        ).toJSON();
        expect(doc).toHaveStyleRule('color', themes.day.fontColor);
        expect(doc).toHaveStyleRule('background-color', themes.day.elementBgColor);
    });

    it('Theme switch', () => {
        act(() => {
            render(<App />);
        });

        const app = document.querySelector('.app');
        const button = document.querySelector('.theme-switch');

        assertStyles(app, {
            color: themes.day.fontColor,
            'background-color': themes.day.mainBgColor
        });
        assertStyles(button, {
            color: themes.day.fontColor,
            'background-color': themes.day.elementBgColor
        });

        user.click(button);

        assertStyles(app, {
            color: themes.night.fontColor,
            'background-color': themes.night.mainBgColor
        });
        assertStyles(button, {
            color: themes.night.fontColor,
            'background-color': themes.night.elementBgColor
        });
    });

    it('StyledButton. Pseudo-elements and pseudo-classes', () => {
        const doc = create(
            <StyledButton theme={themes.day} />
        ).toJSON();
        expect(doc).toHaveStyleRule('background-color', 'red', { modifier: ':hover' });
        expect(doc).toHaveStyleRule('content', '"pops"', { modifier: '::before' });
    });
})
