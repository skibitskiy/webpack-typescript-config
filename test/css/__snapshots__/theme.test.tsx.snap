// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Components/Operation Multiply 1`] = `
.c3 {
  height: 1em;
  width: 2em;
  font-size: 48px;
  color: #000;
  background-color: #e8e8e8;
  border: 2px solid rgba(0,0,0,0.5);
  border-radius: 5px;
  outline: none;
}

.c2 + .c2 {
  margin-left: 16px;
}

.c4 {
  font-size: 48px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 + .c0 {
  margin-top: 24px;
}

<div
  className="c0 c1"
  onChange={[Function]}
>
  <h2>
    Multiply
  </h2>
  <div>
    <input
      className="c2 c3"
      data-index={0}
      value="0"
    />
    <input
      className="c2 c3"
      data-index={1}
      value="0"
    />
  </div>
  <div
    className="c4"
  >
    Ответ: 
    0
  </div>
</div>
`;

exports[`Storyshots Components/Operation Sum 1`] = `
.c3 {
  height: 1em;
  width: 2em;
  font-size: 48px;
  color: #000;
  background-color: #e8e8e8;
  border: 2px solid rgba(0,0,0,0.5);
  border-radius: 5px;
  outline: none;
}

.c2 + .c2 {
  margin-left: 16px;
}

.c4 {
  font-size: 48px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 + .c0 {
  margin-top: 24px;
}

<div
  className="c0 c1"
  onChange={[Function]}
>
  <h2>
    Summarize
  </h2>
  <div>
    <input
      className="c2 c3"
      data-index={0}
      value="0"
    />
    <input
      className="c2 c3"
      data-index={1}
      value="0"
    />
  </div>
  <div
    className="c4"
  >
    Ответ: 
    0
  </div>
</div>
`;
